### ISSUE TITLE
Logging training of TorchModels to WandB

### ISSUE THREAD

**Issue by j-frie** [deepchem/deepchem#2957]
## ❓ Questions & Help

I am trying to log model training of a TorchModel to WandB and have some ideas for improvements:

1. The ValidationCallback is very useful to log metrics for both the validation and the training dataset. Right now the data is logged to wandb using
`data = {'eval/' + k: v for k, v in scores.items()}` and integers can be appended to the key if there are several datasets.
In my opinion, it would be better to have a parameter to control the prefix of the dictionary key when using the ValidationCallback for different datasets. This is what it currently looks like:
![image](https://user-images.githubusercontent.com/107413352/174792587-45905ccd-5f96-477d-a1e3-186f9b88fa97.png)

2. The comments in ValidationCallback are all about Keras models only, but the code seems to work well for PyTorch models. Are there any known issues with PyTorch models or are the comments just outdated?

3. Afaik, the current solutions to get the validation loss are quite hacky (see #1040) and require looping manually over the epochs.  Instead, we could add this functionality to the ValidationCallback (on demand).

I have seen that there are bigger plans to improve the logger integration in general (see #2580), but it looks as if they have been "work in progress" for quite some time...
So I wanted to know your opinions on these small improvements before trying to tackle them and creating a PR.

**Comment by j-frie** [comment]
Thanks for the quick feedback :-)

What is the recommended way to check whether the model that calls the `ValidationCallback` is a `TorchModel` or a `KerasModel`?

The loss calculation is slightly different for the models, but I cannot simply check, if model is an instance of `TorchModel`, because if `torch`is not installed, one cannot import `TorchModel` (and similar for `KerasModel`, if tensorflow is not installed).

The solution I came up with is a bit long, so I was wondering if there is a better way to do it:
```
tf_installed = False
torch_installed = False
try:
  from deepchem.models.keras_model import KerasModel
  tf_installed = True
except ModuleNotFoundError:
  pass
try:
  from deepchem.models.torch_models import TorchModel
  import torch
  torch_installed = True
except ModuleNotFoundError:
  pass
```
The loss calculation itself looks like this:
  ```
def _calculate_loss(self, model):
    if torch_installed and isinstance(model, TorchModel):
      outputs = torch.from_numpy(model.predict(self.dataset))
      labels = torch.from_numpy(self.dataset.y)
      weights = torch.from_numpy(self.dataset.w)
      return = model._loss_fn([outputs], [labels], [weights]).item()

    elif tf_installed and isinstance(model, KerasModel):
      outputs = model.predict(self.dataset)
      labels = self.dataset.y
      weights = self.dataset.w
      return float(model._loss_fn([outputs], [labels], [weights]))
```

**Comment by j-frie** [comment]
That's a good point!

It have now wrapped the loss function (`model._loss_fn`) within the `Model` class to make it usable as a metric function in `Metric`.
```
  def loss_as_metric(self, y_true, y_predicted, sample_weight) -> float:
    """Wrapper to make loss function of Model usable as Metric.

    dc.metrics.Metric expects the metric function to have the args:
    y_true, y_pred, sample_weight
    loss function uses [outputs], [labels], [weights]"""
    loss =  float(self._loss_fn([y_predicted], [y_true], [sample_weight]).numpy())
    return loss
```

In `ValidationCallback` one can then use the existing way to evaluate metrics:
```
loss_metric = Metric(model.loss_as_metric, name='vc_loss', mode='regression')
model.evaluate(self.dataset, [loss_metric], transformers=[], use_sample_weights=True)
```

However, for classification problems, this is a bit tricky, and I need some help:

Some losses like `SoftmaxCrossEntropyLoss` take logits as inputs, but others like `BinaryCrossEntropy` take the output directly. 
How could I get the logits? And how does the model know whether to use logits or outputs for the loss calculation? Is that hard-coded in every single model and we would have to add `evaluate_loss()` to `Model` and override it in all classification models depending on the loss function that is used?
